parlai/core/metrics.py

```diff
@@ -42,6 +47,94 @@
 re_punc = re.compile(r'[!"#$%&()*+,-./:;<=>?@\[\]\\^`{|}~_\']')
 
 
+class Metric(ABC):
+    """
+    Base class for storing metrics.
+
+    Subclasses should define .report().
+    """
+
+    @abstractmethod
+    def report(self) -> float:
+        pass
+
+    def __str__(self) -> str:
+        return f'{self.report():.4g}'
+
+    def __repr__(self) -> str:
+        return f'{self.__class__} ({self.report():.4g})'
+
+
+class SumMetric(Metric):
+    """
+    Class that keeps a running sum of some metric.
+    """
+
+    def __init__(self, sum_: TScalar):
+
+        # Convert sum to float
+        if isinstance(sum_, torch.Tensor):
+            sum_as_number: Union[int, float, bool] = sum_.item()
+        else:
+            sum_as_number = sum_
+        assert isinstance(sum_as_number, Number)
+        sum_as_float = float(sum_as_number)
+
+        self.sum: float = sum_as_float
+
+    def __add__(self, other: 'SumMetric') -> 'SumMetric':
+        # NOTE: hinting can be cleaned up with "from __future__ import annotations" when
+        # we drop Python 3.6
+        full_sum: float = self.sum + other.sum
+        return SumMetric(sum_=full_sum)
+
+    def report(self) -> float:
+        return self.sum
+
+
+class AverageMetric(Metric):
+    """
+    Class that keeps a running average of some metric.
+    """
+
+    def __init__(
+        self, numer: TScalar, denom: TScalar,
+    ):
+
+        # Convert numer to float
+        if isinstance(numer, torch.Tensor):
```

Comment:
looks like we do this "to number" operation many times throughout. can this just be utility function that we call? (perhps with options for int/float)

Reply:
could also be a function of the base `Metric` class